File: samples/swiftui/Sources/DesignSystem/LabeledTextField.swift
import SwiftUI

/// A form field with a mandatory marker and an inline error message.
struct TextFieldModel {
    var label: String
    var value: String
    var isRequired = false
    var error: String?
    var onChange: (String) -> Void = { _ in }
    var onFocusChange: (Bool) -> Void = { _ in }
}

struct LabeledTextField: View {
    let model: TextFieldModel
    @FocusState private var focused: Bool

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            TextField(
                title,
                text: Binding(get: { model.value }, set: model.onChange)
            )
            .textFieldStyle(.roundedBorder)
            .focused($focused)
            .onChange(of: focused) { model.onFocusChange($0) }
            if let error = model.error {
                Text(error)
                    .font(.caption)
                    .foregroundColor(.red)
            }
        }
    }

    private var title: String {
        model.isRequired ? "\(model.label) *" : model.label
    }
}

File: samples/compose/app/src/main/kotlin/dev/getmilano/sample/designsystem/LabeledTextField.kt
package dev.getmilano.sample.designsystem

import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.onFocusChanged

/** A form field with a mandatory marker and an inline error message. */
data class TextFieldModel(
    val label: String,
    val value: String,
    val isRequired: Boolean = false,
    val error: String? = null,
    val onChange: (String) -> Unit = {},
    val onFocusChange: (Boolean) -> Unit = {},
)

@Composable
fun LabeledTextField(model: TextFieldModel) {
    val title = if (model.isRequired) "${model.label} *" else model.label
    OutlinedTextField(
        value = model.value,
        onValueChange = model.onChange,
        label = { Text(title) },
        isError = model.error != null,
        supportingText =
            model.error?.let { error ->
                { Text(error, color = MaterialTheme.colorScheme.error) }
            },
        modifier =
            Modifier
                .fillMaxWidth()
                .onFocusChanged { state -> model.onFocusChange(state.isFocused) },
    )
}
